File: puzzles/php/defibrillators.php
<?php
// Convert degrees to radians
function toRadians($distance) {
  return str_replace(",", ".", $distance) * M_PI / 180;
}

$min = PHP_INT_MAX;
$name = "";
fscanf(STDIN, "%s", $longitude);
fscanf(STDIN, "%s", $latitude);
fscanf(STDIN, "%d", $defibCount);

// Determine the name of the defibrillator located the closest to the user’s position.
for ($i = 0; $i < $defibCount; ++$i) {
  $arr = explode(";", stream_get_line(STDIN, 257, "\n"));
  $defibLon = toRadians($arr[4]);
  $defibLat = toRadians($arr[5]);
  $x = ($defibLon - toRadians($longitude)) * cos((toRadians($latitude) + $defibLat) / 2);
  $y = $defibLat - toRadians($latitude);
  $distance = hypot($x, $y) * 6371;
  if ($distance < $min) {
    $min = $distance;
    $name = $arr[1];
  }
}

echo("$name\n");
?>

File: puzzles/go/defibrillators.go
package main

import (
  "fmt"
  "os"
  "bufio"
  "strings"
  "strconv"
  "math"
)

func main() {
  scanner := bufio.NewScanner(os.Stdin)
  scanner.Buffer(make([]byte, 1000000), 1000000)

  var longitude string
  scanner.Scan()
  fmt.Sscan(scanner.Text(), &longitude)
  var lon = toRadians(longitude)

  var latitude string
  scanner.Scan()
  fmt.Sscan(scanner.Text(), &latitude)
  var lat = toRadians(latitude)
  
  var n int
  scanner.Scan()
  fmt.Sscan(scanner.Text(), &n)
  
  var min = math.Inf(1)
  var name = ""

  for i := 0; i < n; i++ {
    scanner.Scan()
    defib := strings.Split(scanner.Text(), ";")
    var defibLon = toRadians(defib[4])
    var defibLat = toRadians(defib[5])
    var x = (defibLon - lon) * math.Cos((lat + defibLat) / 2)
    var y = defibLat - lat
    var distance = math.Hypot(x, y) * 6371
    if distance < min {
      min = distance
      name = defib[1]
    }
  }
  
  fmt.Println(name)
}

// Convert degrees to radians
func toRadians(angle string) float64 {
  var degrees, _ = strconv.ParseFloat(strings.Replace(angle, ",", ".", -1), 64)
  return degrees * math.Pi / 180
}
